superchain-backend.md

```diff
@@ -0,0 +1,457 @@
+
+## Superchain Backend implementation: `op-resolver`
+
+> *Name suggestions welcome, `op-superchain` seems to generic however.*
+
+This service resolves superchain dependencies, and implements the above proposed spec for a standalone implementation.
+
+A later iteration of this can support the ideated shreds streaming,
+coupled to the block-builder and horizontal tx-pool,
+if and when interop enters super-low-latency territory with intra-block cross-L2 messaging.
+
+Additionally we add a cache for initiating messages, to avoid frequent disk reads of the append-log.
+
+We may be able to shard / distribute the service by isolating the maintenance of each chain append-log,
+and aggregating the verification to handle cyclic-dependencies:
+a service may publish the first unresolved "stuck" block of dependencies to another replica,
+which can then resolve some of them, and forward to the next,
+until all replicas have seen all messages at the "stuck" block height.
+More optimized resolution may be possible, to be explored at a later point.
+Cyclic dependencies are out-of-scope for now:
+sharding without cycle handling should be trivial by attaching superchain-backends of different chains
+to one another and checking message-safety through the standard backend interface.
+
+### Data-model
+
+- Cache for checks on initiating messages
+    - Cache miss:
+        - if tip of chain: check shreds
+        - if within scope: wait for block fetching
+        - if old: fetch individual message
+- Initiating messages and executing messages have a total order within their block (event log index)
+    - needs to be enforced to avoid time-travel hacks
+      (e.g. emit an event that assumes admin permission, and then follow up by giving yourself admin permission)
+    - can maybe be loosened for some specific bridges/assets, to allow flash-loans of these assets
+- Messages are hashed into message-hashes; no point in keeping full message payload around
+- Group messages by their blocks / chains, to not duplicate metadata
+- Warning: executing events can also count as initiating events
+    - We need to specify if we allow an executing message to consume itself,
+      i.e. if initiating is always before executing, if log-index matches.
+    - We can also have cross-chain loops; executing message is used as initiating message and executed by other chain,
+      and executing message on other chain is what is consumed as initiating message by original executing message.
+
+## Embedded RPC-based implementation
+
+As alternative to running a full `op-resolver` service, we may support a fallback in the op-node.
+Message safety is checked by relying on the local execution-engine
+for serving full event-logs and block-safety labels of the local chain,
+and then attaching to another external backend 
+(possibly served by an op-node / execution-engine also) to verify the safety of external messages.
```

Comment:
We can either enforce that additional infra is ran and have all messages be validated or we could "assume true" bitcoin softfork style when the infra isn't set up to do full verification. I am not sure if the assume true style would be able to work in practice due to the way that invalid blocks are mapped to noop blocks - does the noop block end up being batch submitted? I feel like it would need to be batch submitted and the fork choice would need to favor it so that the hash chain in batches can be preserved. If the noop block does get batch submitted, then the "assume true" style syncing should work. Am i thinking about this correctly?

Reply:
The current interop spec assumes that there is a complete available history of cross-L2 data when determining the canonical chain from batch-inputs. If we want to make this not the case, then interop requires deeper batch-submission changes, out of scope for this superchain backend doc. When syncing from a L2 chain directly, without derivation from L1, there will not be an alternative interpretation of inputs with noop-blocks, so then the "assume true" approach can work without superchain backend data (if the safety-level is accepted, but since it's already assuming the chain is canonical, I think this is fine).